std/utf.d

```diff
@@ -3103,17 +3103,33 @@ enum dchar replacementDchar = '\uFFFD';
  * Iterate a range of char, wchar, or dchars by code unit.
  *
  * The purpose is to bypass the special case decoding that
- * $(REF front, std,array) does to character arrays.
+ * $(REF front, std,range,primitives) does to character arrays. As a result,
+ * using ranges with `byCodeUnit` can be `nothrow` while $(REF front, std,range,primitives)
+ * throws when it encounters invalid Unicode sequences.
+ *
+ * A code unit is a building block of the UTF encodings. Generally, an
+ * individual code unit does not represent what's perceived as a full
+ * character. Many characters are encoded with multiple code units. For
+ * example, the UTF-8 code units for `ø` are `0xC3 0xB8`. That
+ * means, an individual element of `byCodeUnit` often does not form a character on
+ * its own. Attempting to treat it as one while iterating over the resulting
+ * range will give nonsensical results.
+ *
  * Params:
- *      r = input range of characters, or array of characters
+ *      r = an input range of characters, or an array of characters
  * Returns:
- *      input range
+ *      An input range with a length if $(XREF traits, isAggregateType)
+ *      is `true` for `R`. Otherwise, this returns a finite random access range
+ *      with slicing.
+ * See_Also:
+ *      Refer to the $(MREF std, uni) docs for a reference on Unicode terminology.
+ *
+ *      If the strings you're dealing with have characters that rely on
+ *      graphemes and cannot be represented in either code points or code
+ *      units, see $(REF byGrapheme, std,uni).
```

Comment:
I think the wording here is wrong, or at least confusing. Every string represents a sequence of graphemes with code units. I think this should be "strings [...] that contain combining characters", or maybe "strings [...] that contain graphemes which are represented/encoded by more that one code unit".

But the presence of such complex graphemes doesn't actually make `byCodeUnit` useless. It just becomes more error-prone. If you know your Unicode and you're really careful, you might be able to squeeze out some nanoseconds by using `byCodeUnit`, only decoding and looking for combining characters when needed.

The paragraph could also give the misleading impression, that you only need `byGrapheme` when dealing with exotic strings, but `byCodeUnit` will be just fine for everyday stuff.

So I suggest to not bind the use of `byGrapheme` to the complexity of the string. Instead just refer to it by what it does:

```
For a range that iterates by grapheme (written character),
see $(REF byGrapheme, std,uni).
```

The warning above about how a code unit is not a character, and the obscure `byCodeUnit` name itself should be enough to make sure that people don't use it accidentally.

(This became an essay. Sorry 'bout that.)